### ISSUE TITLE
Android 9 (API 28) compatibility

### ISSUE THREAD

**Issue by bemasc** [ben-manes/caffeine#293]
Caffeine currently cannot be used in apps that target Android 9 (API 28), because Android 9 blocks access to `threadLocalRandomProbe` as part of the [restrictions on non-SDK interfaces](https://developer.android.com/about/versions/pie/restrictions-non-sdk-interfaces).

For more details see https://github.com/micronaut-projects/micronaut-core/issues/124#issuecomment-415355779

**Comment by ben-manes** [comment]
True, it is lightly touched on in the [Guava compatibility](https://github.com/ben-manes/caffeine/wiki/Guava#android--gwt-compatibility), although the reason has changed now. It is not a goal of this project to run on Android.

However, I think we could provide graceful fallback such as to `ThreadLocal` for broader compatibility. It won't be a goal of this project to target & optimize for Android, though.

**Comment by bemasc** [comment]
Thanks, that sounds like a good plan.

Android does support Java 8, and indeed is running on OpenJDK 8, so perhaps you should revise the rationale there.  Caffeine currently works fine on Android API 24-27, but not 28.
